File: mobile/android/core/src/main/kotlin/com/portalia/core/auth/ResidentSession.kt
package com.portalia.core.auth

import com.portalia.core.api.ResidentApi
import com.portalia.core.model.ResidentMeResponse
import com.portalia.core.net.ApiError
import com.portalia.core.net.HttpClient
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import okhttp3.CookieJar

/**
 * Sessão do app do morador. O refresh token vem no corpo da resposta (não
 * cookie) e é guardado no SecureStore — ver ResidentApi/ResidentAuth no
 * backend, desenhado mobile-first desde o início, ao contrário do fluxo de
 * equipe que foi construído pra navegador primeiro.
 */
class ResidentSession(
    val api: ResidentApi,
    private val store: SecureStore,
) {
    private val _me = MutableStateFlow<ResidentMeResponse?>(null)
    val me: StateFlow<ResidentMeResponse?> = _me

    private val _mustChangePassword = MutableStateFlow(false)
    val mustChangePassword: StateFlow<Boolean> = _mustChangePassword

    private val _isBootstrapping = MutableStateFlow(true)
    val isBootstrapping: StateFlow<Boolean> = _isBootstrapping

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    private var accessToken: String? = null
    private var inFlightRefresh: Deferred<Unit>? = null

    val isAuthenticated: Boolean get() = accessToken != null && _me.value != null

    suspend fun bootstrap() {
        if (store.get(SecureStoreKey.RESIDENT_REFRESH_TOKEN) != null) {
            refresh()
        }
        _isBootstrapping.value = false
    }

    suspend fun login(email: String, password: String): Boolean {
        _errorMessage.value = null
        return try {
            val result = api.login(email, password)
            accessToken = result.accessToken
            store.set(SecureStoreKey.RESIDENT_REFRESH_TOKEN, result.refreshToken)
            _mustChangePassword.value = result.mustChangePassword
            loadMe()
            true
        } catch (e: Exception) {
            _errorMessage.value = (e as? ApiError)?.message ?: "Não foi possível entrar."
            false
        }
    }

    suspend fun logout() {
        accessToken?.let { token -> try { api.logout(token) } catch (e: Exception) { } }
        store.remove(SecureStoreKey.RESIDENT_REFRESH_TOKEN)
        accessToken = null
        _me.value = null
        _mustChangePassword.value = false
    }

    suspend fun changePassword(current: String, new: String): Boolean {
        val token = accessToken ?: return false
        _errorMessage.value = null
        return try {
            val result = api.changePassword(token, current, new)
            accessToken = result.accessToken
            store.set(SecureStoreKey.RESIDENT_REFRESH_TOKEN, result.refreshToken)
            _mustChangePassword.value = false
            true
        } catch (e: Exception) {
            _errorMessage.value = (e as? ApiError)?.message ?: "Não foi possível trocar a senha."
            false
        }
    }

    suspend fun <T> authorized(call: suspend (String) -> T): T {
        val token = accessToken ?: throw ApiError.Unauthorized
        return try {
            call(token)
        } catch (e: ApiError.Unauthorized) {
            refresh()
            val retryToken = accessToken ?: throw ApiError.Unauthorized
            call(retryToken)
        }
    }

    private suspend fun refresh() {
        val existing = inFlightRefresh
        if (existing != null) {
            existing.await()
            return
        }
        coroutineScope {
            val task = async { performRefresh() }
            inFlightRefresh = task
            task.await()
            inFlightRefresh = null
        }
    }

    private suspend fun performRefresh() {
        val refreshToken = store.get(SecureStoreKey.RESIDENT_REFRESH_TOKEN)
        if (refreshToken == null) {
            accessToken = null
            _me.value = null
            return
        }
        try {
            val result = api.refresh(refreshToken)
            accessToken = result.accessToken
            store.set(SecureStoreKey.RESIDENT_REFRESH_TOKEN, result.refreshToken)
            _mustChangePassword.value = result.mustChangePassword
            if (_me.value == null) loadMe()
        } catch (e: Exception) {
            store.remove(SecureStoreKey.RESIDENT_REFRESH_TOKEN)
            accessToken = null
            _me.value = null
        }
    }

    private suspend fun loadMe() {
        val token = accessToken ?: return
        try {
            _me.value = api.me(token)
        } catch (e: Exception) {
            _errorMessage.value = (e as? ApiError)?.message
        }
    }

    companion object {
        fun create(context: android.content.Context): ResidentSession {
            val store = SecureStore.shared(context)
            val http = HttpClient(cookieJar = CookieJar.NO_COOKIES)
            return ResidentSession(ResidentApi(http), store)
        }
    }
}

File: mobile/ios/PortaliaCore/Sources/PortaliaCore/Auth/ResidentSession.swift
import Foundation

/// Sessão do app do morador. Ao contrário da equipe, o refresh token não vem
/// de um cookie — o backend devolve o valor no corpo e o app guarda no
/// Keychain, reapresentando explicitamente em `POST /resident-auth/refresh`.
/// Isso é o que permite reabrir o app depois de fechado sem precisar logar
/// de novo (o Keychain sobrevive o processo morrer, memória não).
@MainActor
public final class ResidentSession: ObservableObject {
    @Published public private(set) var me: ResidentMeResponse?
    @Published public private(set) var mustChangePassword = false
    @Published public private(set) var isBootstrapping = true
    @Published public var errorMessage: String?

    public let api: ResidentAPI
    private let keychain: KeychainStore
    private var accessToken: String?
    private var inFlightRefresh: Task<Void, Never>?

    public var isAuthenticated: Bool { accessToken != nil && me != nil }

    public init(api: ResidentAPI = ResidentAPI(http: HTTPClient()), keychain: KeychainStore = .shared) {
        self.api = api
        self.keychain = keychain
    }

    public func bootstrap() async {
        if keychain.get(KeychainKey.residentRefreshToken) != nil {
            await refresh()
        }
        isBootstrapping = false
    }

    public func login(email: String, password: String) async -> Bool {
        errorMessage = nil
        do {
            let result = try await api.login(email: email, password: password)
            accessToken = result.accessToken
            keychain.set(result.refreshToken, for: KeychainKey.residentRefreshToken)
            mustChangePassword = result.mustChangePassword
            await loadMe()
            return true
        } catch {
            errorMessage = (error as? APIError)?.errorDescription ?? "Não foi possível entrar."
            return false
        }
    }

    public func logout() async {
        if let accessToken {
            try? await api.logout(accessToken: accessToken)
        }
        keychain.remove(KeychainKey.residentRefreshToken)
        self.accessToken = nil
        self.me = nil
        self.mustChangePassword = false
    }

    public func changePassword(current: String, new: String) async -> Bool {
        guard let accessToken else { return false }
        errorMessage = nil
        do {
            let result = try await api.changePassword(accessToken: accessToken, current: current, new: new)
            self.accessToken = result.accessToken
            keychain.set(result.refreshToken, for: KeychainKey.residentRefreshToken)
            self.mustChangePassword = false
            return true
        } catch {
            errorMessage = (error as? APIError)?.errorDescription ?? "Não foi possível trocar a senha."
            return false
        }
    }

    public func authorized<T>(_ call: (String) async throws -> T) async throws -> T {
        guard let token = accessToken else { throw APIError.unauthorized }
        do {
            return try await call(token)
        } catch APIError.unauthorized {
            await refresh()
            guard let retryToken = accessToken else { throw APIError.unauthorized }
            return try await call(retryToken)
        }
    }

    /// Exposto para telas de SIP que precisam do access token bruto (ex.: chamar
    /// GET /resident/telephony/credentials fora do wrapper `authorized`).
    public func currentAccessToken() -> String? { accessToken }

    private func refresh() async {
        if let inFlightRefresh {
            await inFlightRefresh.value
            return
        }
        // Ver a mesma anotação em StaffSession: sem ela o Swift infere
        // `Task<Void?, Never>` por causa do opcional encadeado em `self?`.
        let task: Task<Void, Never> = Task { [weak self] in await self?.performRefresh() }
        inFlightRefresh = task
        await task.value
        inFlightRefresh = nil
    }

    private func performRefresh() async {
        guard let refreshToken = keychain.get(KeychainKey.residentRefreshToken) else {
            accessToken = nil
            me = nil
            return
        }
        do {
            let result = try await api.refresh(refreshToken: refreshToken)
            self.accessToken = result.accessToken
            keychain.set(result.refreshToken, for: KeychainKey.residentRefreshToken)
            self.mustChangePassword = result.mustChangePassword
            if me == nil { await loadMe() }
        } catch {
            keychain.remove(KeychainKey.residentRefreshToken)
            self.accessToken = nil
            self.me = nil
        }
    }

    private func loadMe() async {
        guard let accessToken else { return }
        do {
            me = try await api.me(accessToken: accessToken)
        } catch {
            errorMessage = (error as? APIError)?.errorDescription
        }
    }
}
